animate/metric.py

```diff
@@ -181,13 +232,14 @@ def copy(self, deepcopy=False):
         """
         Copy the metric and any associated parameters.
 
-        :kwarg deepcopy: If ``True``, the new :class:`~.RiemannianMetric` will allocate
-            new space and copy values.  If ``False``, the default, then the new
-            :class:`~.RiemannianMetric` will share the dof values.
+        :kwarg deepcopy: If ``True``, the new metric will allocate new space and copy
+            values. If ``False`` (default) then the new metric will share the DoF values.
+        :type deepcopy: :class:`bool`
         :return: a copy of the metric with the same parameters set
+        :rtype: :class:`~.RiemannianMetric`
         """
         metric = type(self)(super().copy(deepcopy=deepcopy))
-        metric.set_parameters(self.metric_parameters.copy())
+        metric.set_parameters(self.metric_parameters)
```

Comment:
I see that you put in a unit test for this which passes for me. But when I do the following:
```
from firedrake import *
from animate.metric import RiemannianMetric


mesh = UnitSquareMesh(1, 1)
metric1 = RiemannianMetric(TensorFunctionSpace(mesh, "CG", 1))

P1 = FunctionSpace(mesh, "CG", 1)
mp1 = {"h_min": Function(P1).assign(1.0), "h_max": 1.0}
metric1.set_parameters(mp1)

metric2 = metric1.copy(deepcopy=True)
mp2 = metric2.metric_parameters

print(id(mp1) == id(mp2))                    # False
print(id(mp1["h_min"]) == id(mp2["h_min"]))  # True
print(id(mp1["h_max"]) == id(mp2["h_max"]))  # True

mp1["h_min"].assign(999)
mp1["h_max"] = 999

print(mp1["h_min"].dat.data)   # 999
print(mp1["h_max"])            # 999
print("-------------------")
print(mp2["h_min"].dat.data)  # 999  - WRONG
print(mp2["h_max"])           # 1.0

print(id(mp1) == id(mp2))                    # False
print(id(mp1["h_min"]) == id(mp2["h_min"]))  # True
print(id(mp1["h_max"]) == id(mp2["h_max"]))  # False
```

it doesn't seem to deep-copy variable parameters. I thought it might make a difference if we define the `Function` outside of the dictionary, i.e. sth like
```
h_min = Function(...)
mp = {"h_min": h_min}
```
but that doesn't seem to make a difference. So I'm not sure why the test passes and this doesn't... I don't see a difference

Reply:
Ah I see... It's because in this example I lazily wrote `"h_min"` instead of "`dm_plex_metric_h_min"` (which I always do).

Now I feel like having two accepted ways of naming parameters is just increasing chances for issues popping up. Why not just stick with one to make sure the code is bulletproof?

Reply:
I'm ignoring the first comment based on our discussion on the original issue. See 3ae4d9e900362bd0adb22db1848c1623c6b769d6 for my addressing your latest comment.